core/description/resource.go

```diff
@@ -0,0 +1,317 @@
+	checker := versionedChecker("resources")
+	coerced, err := checker.Coerce(source, nil)
+	if err != nil {
+		return nil, errors.Annotate(err, "resources version schema check failed")
+	}
+	valid := coerced.(map[string]interface{})
+
+	version := int(valid["version"].(int64))
+	importFunc, ok := resourceDeserializationFuncs[version]
+	if !ok {
+		return nil, errors.NotValidf("version %d", version)
+	}
+	sourceList := valid["resources"].([]interface{})
+	return importResourceList(sourceList, importFunc)
+}
+
+func importResourceList(sourceList []interface{}, importFunc resourceDeserializationFunc) ([]*resource, error) {
+	result := make([]*resource, 0, len(sourceList))
+	for i, value := range sourceList {
+		source, ok := value.(map[string]interface{})
+		if !ok {
+			return nil, errors.Errorf("unexpected value for resource %d, %T", i, value)
+		}
+		resource, err := importFunc(source)
+		if err != nil {
+			return nil, errors.Annotatef(err, "resource %d", i)
+		}
+		result = append(result, resource)
+	}
+	return result, nil
+}
+
+type resourceDeserializationFunc func(map[string]interface{}) (*resource, error)
+
+var resourceDeserializationFuncs = map[int]resourceDeserializationFunc{
+	1: importResourceV1,
+}
+
+func importResourceV1(source map[string]interface{}) (*resource, error) {
+	fields := schema.Fields{
+		"name":                 schema.String(),
+		"application-revision": schema.Int(),
+		"charmstore-revision":  schema.Int(),
+		"revisions":            schema.List(schema.StringMap(schema.Any())),
+	}
+	checker := schema.FieldMap(fields, nil)
+
+	coerced, err := checker.Coerce(source, nil)
+	if err != nil {
+		return nil, errors.Annotatef(err, "resource v1 schema check failed")
+	}
+	valid := coerced.(map[string]interface{})
+
+	// From here we know that the map returned from the schema coercion
+	// contains fields of the right type.
+	name := valid["name"].(string)
+	revList := valid["revisions"].([]interface{})
+	r := newResource(ResourceArgs{
+		Name:               name,
+		Revision:           int(valid["application-revision"].(int64)),
+		CharmStoreRevision: int(valid["charmstore-revision"].(int64)),
+	})
+	for i, revSource := range revList {
+		revision, err := importResourceRevisionV1(revSource)
+		if err != nil {
+			return nil, errors.Annotatef(err, "resource %s, revision index %d", name, i)
+		}
+		r.Revisions_ = append(r.Revisions_, revision)
+	}
+	return r, nil
+}
+
+func importResourceRevisionV1(source interface{}) (*resourceRevision, error) {
+	fields := schema.Fields{
+		"revision":      schema.Int(),
+		"type":          schema.String(),
+		"path":          schema.String(),
+		"description":   schema.String(),
+		"origin":        schema.String(),
+		"fingerprint":   schema.String(),
+		"size":          schema.Int(),
+		"add-timestamp": schema.Time(),
+		"username":      schema.String(),
+	}
+	defaults := schema.Defaults{
+		"add-timestamp": time.Time{},
```

Comment:
It would be easier to have schema.Omit for the timestamp, and then only set it below if it exists, rather than go through zero.